File: opencode_profiles/ops.py
import json
import os
import shutil
from datetime import datetime

from opencode_profiles.paths import OpenCodePaths
from opencode_profiles.skills import scan_current_skills, write_skills_yml


def ensure_initialized(paths: OpenCodePaths) -> None:
    """确保 opencode 配置目录已初始化。

    如果 opencode.json 不是 symlink，将其内容存入 default profile 并替换为 symlink。
    如果已经是有效 symlink（目标存在），不做任何操作。
    如果 symlink 悬空（目标不存在），清理后重新初始化。
    """
    paths.profiles_dir.mkdir(parents=True, exist_ok=True)

    config = paths.config_file

    # Valid symlink: target exists
    if config.is_symlink() and config.exists():
        # Ensure default profile has skills.yml even on existing setups
        if not paths.profile_skills_yml("default").exists():
            current = scan_current_skills(paths)
            write_skills_yml(paths, "default", current)
        return

    # Clean up dangling symlink
    if config.is_symlink():
        config.unlink()

    default_dir = paths.profile_dir("default")
    default_dir.mkdir(parents=True, exist_ok=True)
    default_config = paths.profile_config("default")
    default_tui_config = paths.profile_tui_config("default")
    skills_dir = paths.profile_skills("default")
    skills_dir.mkdir(exist_ok=True)

    if config.exists():
        backup_path = paths.base_dir / "opencode.json.bak"
        if not backup_path.exists():
            shutil.copy2(config, backup_path)
        shutil.copy2(config, default_config)
        config.unlink()
    else:
        # Config doesn't exist (dangling symlink removed or fresh install)
        # Restore from backup if available
        backup_path = paths.base_dir / "opencode.json.bak"
        if backup_path.exists():
            shutil.copy2(backup_path, default_config)
        else:
            default_config.write_text("{}")

    os.symlink(paths.relative_target("default"), config)

    # Handle tui.json (also check for dangling symlink)
    tui_config = paths.tui_config_file
    if tui_config.is_symlink() and not tui_config.exists():
        tui_config.unlink()

    if tui_config.exists() and not tui_config.is_symlink():
        shutil.copy2(tui_config, default_tui_config)
        tui_config.unlink()
        os.symlink(paths.relative_target_tui("default"), tui_config)

    # Write skills.yml for default profile
    if not paths.profile_skills_yml("default").exists():
        current = scan_current_skills(paths)
        write_skills_yml(paths, "default", current)


def backup(paths: OpenCodePaths) -> str:
    """备份当前激活的 profile 配置。返回备份目录名称。"""
    ensure_initialized(paths)

    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    backup_name = f"backup_{timestamp}"
    backup_dir = paths.profile_dir(backup_name)
    backup_dir.mkdir(parents=True, exist_ok=True)

    current_config = paths.config_file
    if current_config.is_symlink():
        target = current_config.resolve()
        shutil.copy2(target, paths.profile_config(backup_name))
    else:
        raise RuntimeError("config_file is not a symlink after init")

    current_tui = paths.tui_config_file
    if current_tui.is_symlink():
        tui_target = current_tui.resolve()
        shutil.copy2(tui_target, paths.profile_tui_config(backup_name))

    paths.profile_skills(backup_name).mkdir(exist_ok=True)
    return backup_name


def create_from_current(paths: OpenCodePaths, name: str) -> None:
    """从当前激活的 profile 创建新 profile。"""
    ensure_initialized(paths)

    current_config = paths.config_file
    if not current_config.is_symlink():
        raise RuntimeError("config_file is not a symlink after init")
    target = current_config.resolve()

    profile_dir = paths.profile_dir(name)
    if profile_dir.exists():
        # 先复制到临时位置（保持元数据），再 rmtree，再 move 回来——
        # 因为当覆盖当前激活的 profile 时，源和目标是同一个文件，
        # 必须先保存内容再删除目录。
        tmp_path = paths.base_dir / f".opencode.json.tmp.{name}"
        shutil.copy2(target, tmp_path)
        shutil.rmtree(profile_dir)
        profile_dir.mkdir(parents=True)
        paths.profile_skills(name).mkdir(exist_ok=True)
        shutil.move(tmp_path, paths.profile_config(name))
    else:
        profile_dir.mkdir(parents=True)
        paths.profile_skills(name).mkdir(exist_ok=True)
        shutil.copy2(target, paths.profile_config(name))

    active = get_active(paths)
    if active is not None:
        current_tui = paths.profile_tui_config(active)
        if current_tui.exists():
            shutil.copy2(current_tui, paths.profile_tui_config(name))

    # Write skills.yml for new profile
    current = scan_current_skills(paths)
    write_skills_yml(paths, name, current)


def _load_providers(paths: OpenCodePaths, source: str) -> dict:
    """从源配置读取 provider dict。source 为 'current' 或 profile 名。

    Raises:
        FileNotFoundError: 源配置不存在
        ValueError: 源配置无 provider 或 provider 为空
    """
    if source == "current":
        config = paths.config_file
        if not config.is_symlink():
            raise FileNotFoundError("Current config is not a symlink")
        target = config.resolve()
        data = json.loads(target.read_text())
    else:
        config_path = paths.profile_config(source)
        if not config_path.exists():
            raise FileNotFoundError(f"Source profile '{source}' not found")
        data = json.loads(config_path.read_text())

    providers = data.get("provider")
    if not providers:
        raise ValueError("Source config has no providers to import")
    return providers


def create_empty(paths: OpenCodePaths, name: str, source: str | None = None) -> None:
    """创建空 profile，可选从源配置导入 provider。

    Args:
        paths: 路径管理实例
        name: 新 profile 名称
        source: None 表示空配置；"current" 表示从当前激活配置导入；
                其他字符串表示从指定 profile 名称导入
    """
    ensure_initialized(paths)

    profile_dir = paths.profile_dir(name)
    if profile_dir.exists():
        shutil.rmtree(profile_dir)

    providers = _load_providers(paths, source) if source is not None else None

    profile_dir.mkdir(parents=True)
    paths.profile_skills(name).mkdir(exist_ok=True)

    if providers is None:
        paths.profile_config(name).write_text("{}")
    else:
        paths.profile_config(name).write_text(json.dumps({"provider": providers}, indent=2))

    if source is not None:
        if source == "current":
            active = get_active(paths)
            if active is None:
                # Still write skills.yml before returning
                current = scan_current_skills(paths)
                write_skills_yml(paths, name, current)
                return
            src_tui = paths.profile_tui_config(active)
        else:
            src_tui = paths.profile_tui_config(source)
        if src_tui.exists():
            shutil.copy2(src_tui, paths.profile_tui_config(name))

    # Write skills.yml for new profile
    current = scan_current_skills(paths)
    write_skills_yml(paths, name, current)


def switch(paths: OpenCodePaths, name: str) -> None:
    """切换 symlink 指向目标 profile。"""
    ensure_initialized(paths)

    target = paths.profile_config(name)
    if not target.exists():
        available = list_profiles(paths)
        raise FileNotFoundError(f"Profile '{name}' not found. Available: {available}")

    config = paths.config_file
    config.parent.mkdir(parents=True, exist_ok=True)

    if config.is_symlink() or config.exists():
        config.unlink()

    os.symlink(paths.relative_target(name), config)

    # 管理 tui.json symlink
    tui_config = paths.tui_config_file
    tui_target = paths.profile_tui_config(name)

    if tui_target.exists():
        if tui_config.is_symlink() or tui_config.exists():
            tui_config.unlink()
        os.symlink(paths.relative_target_tui(name), tui_config)
    elif tui_config.is_symlink():
        tui_config.unlink()

    # Sync skills symlinks
    from opencode_profiles.skills import sync_skills

    sync_skills(paths, name)


def list_profiles(paths: OpenCodePaths) -> list[str]:
    """列出所有 profile 名称。"""
    if not paths.profiles_dir.exists():
        return []

    profiles = []
    for d in sorted(paths.profiles_dir.iterdir()):
        if d.is_dir() and (d / "opencode.json").exists():
            profiles.append(d.name)
    return profiles


def get_active(paths: OpenCodePaths) -> str | None:
    """获取当前激活的 profile 名称。"""
    config = paths.config_file
    if not config.is_symlink():
        return None

    target = config.resolve()
    parts = target.relative_to(paths.base_dir).parts
    if len(parts) >= 2 and parts[0] == "profiles":
        return parts[1]
    return None

File: internal/ops/ops.go
package ops

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"opencode-profiles/internal/paths"
	"opencode-profiles/internal/skills"
)

// copyFile 复制文件内容、权限与时间戳（近似 shutil.copy2）。
func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	fi, err := in.Stat()
	if err != nil {
		return err
	}
	out, err := os.OpenFile(dst, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fi.Mode())
	if err != nil {
		return err
	}
	if _, err := io.Copy(out, in); err != nil {
		out.Close()
		return err
	}
	if err := out.Close(); err != nil {
		return err
	}
	return os.Chtimes(dst, fi.ModTime(), fi.ModTime())
}

func isSymlink(path string) bool {
	fi, err := os.Lstat(path)
	return err == nil && fi.Mode()&os.ModeSymlink != 0
}

// pythonList 将 Go 切片渲染为 Python repr 风格的列表（用于错误消息）。
func pythonList(items []string) string {
	quoted := make([]string, len(items))
	for i, s := range items {
		quoted[i] = "'" + s + "'"
	}
	return "[" + strings.Join(quoted, ", ") + "]"
}

// EnsureInitialized 确保 opencode 配置目录已初始化。
//
// 如果 opencode.json 不是 symlink，将其内容存入 default profile 并替换为 symlink。
// 如果已经是有效 symlink（目标存在），不做任何操作。
// 如果 symlink 悬空（目标不存在），清理后重新初始化。
func EnsureInitialized(p *paths.Paths) error {
	if err := os.MkdirAll(p.ProfilesDir(), 0o755); err != nil {
		return err
	}

	config := p.ConfigFile()

	// Valid symlink: target exists
	if isSymlink(config) {
		if _, err := os.Stat(config); err == nil {
			if _, err := os.Stat(p.ProfileSkillsYML("default")); os.IsNotExist(err) {
				current, err := skills.ScanCurrentSkills(p)
				if err != nil {
					return err
				}
				if err := skills.WriteSkillsYML(p, "default", current); err != nil {
					return err
				}
			}
			return nil
		}
	}

	// Clean up dangling symlink
	if isSymlink(config) {
		if err := os.Remove(config); err != nil {
			return err
		}
	}

	defaultDir := p.ProfileDir("default")
	if err := os.MkdirAll(defaultDir, 0o755); err != nil {
		return err
	}
	defaultConfig := p.ProfileConfig("default")
	defaultTUIConfig := p.ProfileTUIConfig("default")
	skillsDir := p.ProfileSkills("default")
	if err := os.MkdirAll(skillsDir, 0o755); err != nil {
		return err
	}

	if _, err := os.Stat(config); err == nil {
		backupPath := filepath.Join(p.BaseDir(), "opencode.json.bak")
		if _, err := os.Stat(backupPath); os.IsNotExist(err) {
			if err := copyFile(config, backupPath); err != nil {
				return err
			}
		}
		if err := copyFile(config, defaultConfig); err != nil {
			return err
		}
		if err := os.Remove(config); err != nil {
			return err
		}
	} else {
		// Config doesn't exist (dangling symlink removed or fresh install)
		backupPath := filepath.Join(p.BaseDir(), "opencode.json.bak")
		if _, err := os.Stat(backupPath); err == nil {
			if err := copyFile(backupPath, defaultConfig); err != nil {
				return err
			}
		} else {
			if err := os.WriteFile(defaultConfig, []byte("{}"), 0o644); err != nil {
				return err
			}
		}
	}

	if err := os.Symlink(p.RelativeTarget("default"), config); err != nil {
		return err
	}

	// Handle tui.json (also check for dangling symlink)
	tuiConfig := p.TUIConfigFile()
	tuiWasDangling := isSymlink(tuiConfig)
	if tuiWasDangling {
		if _, err := os.Stat(tuiConfig); err == nil {
			tuiWasDangling = false
		}
	}
	if tuiWasDangling {
		if err := os.Remove(tuiConfig); err != nil {
			return err
		}
	}

	if _, err := os.Stat(tuiConfig); err == nil && !isSymlink(tuiConfig) {
		if err := copyFile(tuiConfig, defaultTUIConfig); err != nil {
			return err
		}
		if err := os.Remove(tuiConfig); err != nil {
			return err
		}
		if err := os.Symlink(p.RelativeTargetTUI("default"), tuiConfig); err != nil {
			return err
		}
	} else if tuiWasDangling {
		// tui.json was a dangling symlink — recreate symlink pointing
		// to default profile; create empty tui.json in default if needed.
		if _, err := os.Stat(defaultTUIConfig); os.IsNotExist(err) {
			if err := os.WriteFile(defaultTUIConfig, []byte("{}"), 0o644); err != nil {
				return err
			}
		}
		if err := os.Symlink(p.RelativeTargetTUI("default"), tuiConfig); err != nil {
			return err
		}
	}

	// Write skills.yml for default profile
	if _, err := os.Stat(p.ProfileSkillsYML("default")); os.IsNotExist(err) {
		current, err := skills.ScanCurrentSkills(p)
		if err != nil {
			return err
		}
		if err := skills.WriteSkillsYML(p, "default", current); err != nil {
			return err
		}
	}

	return nil
}

// Backup 备份当前激活的 profile 配置。返回备份目录名称。
func Backup(p *paths.Paths) (string, error) {
	if err := EnsureInitialized(p); err != nil {
		return "", err
	}

	timestamp := time.Now().Format("20060102_150405")
	backupName := "backup_" + timestamp
	backupDir := p.ProfileDir(backupName)
	if err := os.MkdirAll(backupDir, 0o755); err != nil {
		return "", err
	}

	currentConfig := p.ConfigFile()
	if !isSymlink(currentConfig) {
		return "", fmt.Errorf("config_file is not a symlink after init")
	}
	target, err := filepath.EvalSymlinks(currentConfig)
	if err != nil {
		return "", err
	}
	if err := copyFile(target, p.ProfileConfig(backupName)); err != nil {
		return "", err
	}

	currentTUI := p.TUIConfigFile()
	if isSymlink(currentTUI) {
		tuiTarget, err := filepath.EvalSymlinks(currentTUI)
		if err != nil {
			return "", err
		}
		if err := copyFile(tuiTarget, p.ProfileTUIConfig(backupName)); err != nil {
			return "", err
		}
	}

	if err := os.MkdirAll(p.ProfileSkills(backupName), 0o755); err != nil {
		return "", err
	}
	return backupName, nil
}

// CreateFromCurrent 从当前激活的 profile 创建新 profile。
func CreateFromCurrent(p *paths.Paths, name string) error {
	if err := EnsureInitialized(p); err != nil {
		return err
	}

	currentConfig := p.ConfigFile()
	if !isSymlink(currentConfig) {
		return fmt.Errorf("config_file is not a symlink after init")
	}
	target, err := filepath.EvalSymlinks(currentConfig)
	if err != nil {
		return err
	}

	// 读取 tui.json 内容（在 rmtree 之前，防止覆盖激活 profile 时源文件被删）
	var tuiContent *string
	if active := GetActive(p); active != "" {
		currentTUI := p.ProfileTUIConfig(active)
		if data, err := os.ReadFile(currentTUI); err == nil {
			s := string(data)
			tuiContent = &s
		}
	}

	profileDir := p.ProfileDir(name)
	if _, err := os.Stat(profileDir); err == nil {
		// 先复制到临时位置（保持元数据），再 rmtree，再 move 回来——
		// 因为当覆盖当前激活的 profile 时，源和目标是同一个文件，
		// 必须先保存内容再删除目录。
		tmpPath := filepath.Join(p.BaseDir(), ".opencode.json.tmp."+name)
		if err := copyFile(target, tmpPath); err != nil {
			return err
		}
		if err := os.RemoveAll(profileDir); err != nil {
			_ = os.Remove(tmpPath)
			return err
		}
		if err := os.MkdirAll(profileDir, 0o755); err != nil {
			_ = os.Remove(tmpPath)
			return err
		}
		if err := os.MkdirAll(p.ProfileSkills(name), 0o755); err != nil {
			_ = os.Remove(tmpPath)
			return err
		}
		if err := os.Rename(tmpPath, p.ProfileConfig(name)); err != nil {
			_ = os.Remove(tmpPath)
			return err
		}
	} else if os.IsNotExist(err) {
		if err := os.MkdirAll(profileDir, 0o755); err != nil {
			return err
		}
		if err := os.MkdirAll(p.ProfileSkills(name), 0o755); err != nil {
			return err
		}
		if err := copyFile(target, p.ProfileConfig(name)); err != nil {
			return err
		}
	} else {
		return err
	}

	// 写入 tui.json（使用提前读取的内容）
	if tuiContent != nil {
		if err := os.WriteFile(p.ProfileTUIConfig(name), []byte(*tuiContent), 0o644); err != nil {
			return err
		}
	}

	// Write skills.yml for new profile
	current, err := skills.ScanCurrentSkills(p)
	if err != nil {
		return err
	}
	return skills.WriteSkillsYML(p, name, current)
}

// LoadProviders 从源配置读取 provider dict。source 为 "current" 或 profile 名。
func LoadProviders(p *paths.Paths, source string) (map[string]any, error) {
	var data map[string]any
	if source == "current" {
		config := p.ConfigFile()
		if !isSymlink(config) {
			return nil, fmt.Errorf("Current config is not a symlink")
		}
		target, err := filepath.EvalSymlinks(config)
		if err != nil {
			return nil, err
		}
		raw, err := os.ReadFile(target)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal(raw, &data); err != nil {
			return nil, err
		}
	} else {
		configPath := p.ProfileConfig(source)
		if _, err := os.Stat(configPath); err != nil {
			if os.IsNotExist(err) {
				return nil, fmt.Errorf("Source profile '%s' not found", source)
			}
			return nil, err
		}
		raw, err := os.ReadFile(configPath)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal(raw, &data); err != nil {
			return nil, err
		}
	}

	providers, _ := data["provider"].(map[string]any)
	if len(providers) == 0 {
		return nil, fmt.Errorf("Source config has no providers to import")
	}
	return providers, nil
}

// CreateEmpty 创建空 profile，可选从源配置导入 provider。
// source 为空字符串表示空配置；"current" 表示从当前激活配置导入；
// 其他字符串表示从指定 profile 名称导入。
func CreateEmpty(p *paths.Paths, name, source string) error {
	if err := EnsureInitialized(p); err != nil {
		return err
	}

	profileDir := p.ProfileDir(name)
	if _, err := os.Stat(profileDir); err == nil {
		if err := os.RemoveAll(profileDir); err != nil {
			return err
		}
	}

	var providers map[string]any
	if source != "" {
		var err error
		providers, err = LoadProviders(p, source)
		if err != nil {
			return err
		}
	}

	if err := os.MkdirAll(profileDir, 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(p.ProfileSkills(name), 0o755); err != nil {
		return err
	}

	if providers == nil {
		return os.WriteFile(p.ProfileConfig(name), []byte("{}"), 0o644)
	}
	data, err := json.MarshalIndent(map[string]any{"provider": providers}, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(p.ProfileConfig(name), data, 0o644)
}

// SwitchDB 切换 symlink 指向目标 profile 并同步技能（dbPath 注入用）。
func SwitchDB(p *paths.Paths, name, dbPath string) error {
	if err := EnsureInitialized(p); err != nil {
		return err
	}

	target := p.ProfileConfig(name)
	if _, err := os.Stat(target); err != nil {
		if os.IsNotExist(err) {
			available := ListProfiles(p)
			return fmt.Errorf("Profile '%s' not found. Available: %s", name, pythonList(available))
		}
		return err
	}

	config := p.ConfigFile()
	if err := os.MkdirAll(filepath.Dir(config), 0o755); err != nil {
		return err
	}

	if isSymlink(config) {
		if err := os.Remove(config); err != nil {
			return err
		}
	} else if _, err := os.Stat(config); err == nil {
		if err := os.Remove(config); err != nil {
			return err
		}
	}

	if err := os.Symlink(p.RelativeTarget(name), config); err != nil {
		return err
	}

	// 管理 tui.json symlink
	tuiConfig := p.TUIConfigFile()
	tuiTarget := p.ProfileTUIConfig(name)

	if _, err := os.Stat(tuiTarget); err == nil {
		if isSymlink(tuiConfig) {
			if err := os.Remove(tuiConfig); err != nil {
				return err
			}
		} else if _, err := os.Stat(tuiConfig); err == nil {
			if err := os.Remove(tuiConfig); err != nil {
				return err
			}
		}
		if err := os.Symlink(p.RelativeTargetTUI(name), tuiConfig); err != nil {
			return err
		}
	} else if isSymlink(tuiConfig) {
		if err := os.Remove(tuiConfig); err != nil {
			return err
		}
	}

	// Sync skills symlinks
	return skills.SyncSkills(p, name, dbPath)
}

// Switch 切换 symlink 指向目标 profile（使用默认 db 路径）。
func Switch(p *paths.Paths, name string) error {
	return SwitchDB(p, name, "")
}

// ListProfiles 列出所有 profile 名称。
func ListProfiles(p *paths.Paths) []string {
	entries, err := os.ReadDir(p.ProfilesDir())
	if err != nil {
		return nil
	}
	var profiles []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		configPath := filepath.Join(p.ProfileDir(entry.Name()), "opencode.json")
		if _, err := os.Stat(configPath); err == nil {
			profiles = append(profiles, entry.Name())
		}
	}
	sort.Strings(profiles)
	return profiles
}

// GetActive 获取当前激活的 profile 名称；非 symlink 时返回空字符串。
func GetActive(p *paths.Paths) string {
	config := p.ConfigFile()
	if !isSymlink(config) {
		return ""
	}

	target, err := filepath.EvalSymlinks(config)
	if err != nil {
		return ""
	}
	rel, err := filepath.Rel(p.BaseDir(), target)
	if err != nil {
		return ""
	}
	parts := strings.Split(rel, string(filepath.Separator))
	if len(parts) >= 2 && parts[0] == "profiles" {
		return parts[1]
	}
	return ""
}
